### PR TITLE
backend/enhancement/sync

### PR DESCRIPTION
**Pull request by max-keviv**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
pass the validation for cronjob sync api
added rideEndedBy in ride provider platform 
cronjob validation is removed in the following validation handler
validateRideCompletedReq and validateBookingCancelledReq

### CODE REVIEW

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/Common.hs [hunk 1]
```diff
@@ -478,8 +478,9 @@ validateRideAssignedReq ::
   RideAssignedReq ->
   m ValidatedRideAssignedReq
 validateRideAssignedReq RideAssignedReq {..} = do
+  let BookingDetails {..} = bookingDetails
   booking <- QRB.findByTransactionId transactionId >>= fromMaybeM (BookingDoesNotExist $ "transactionId:-" <> transactionId)
-  unless (isAssignable booking) $ throwError (BookingInvalidStatus $ show booking.status)
+  unless (isAssignable booking || isInitiatedByCronJob) $ throwError (BookingInvalidStatus $ show booking.status)
```
Comment (jaypalmudaliyar24) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/Common.hs:483]: Not required for `RIDE_ASSIGNED`.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/Common.hs [hunk 2]
```diff
@@ -498,8 +499,8 @@ validateRideStartedReq RideStartedReq {..} = do
   let BookingDetails {..} = bookingDetails
   booking <- QRB.findByBPPBookingId bppBookingId >>= fromMaybeM (BookingDoesNotExist $ "BppBookingId: " <> bppBookingId.getId)
   ride <- QRide.findByBPPRideId bppRideId >>= fromMaybeM (RideDoesNotExist $ "BppRideId" <> bppRideId.getId)
-  unless (booking.status == DRB.TRIP_ASSIGNED) $ throwError (BookingInvalidStatus $ show booking.status)
-  unless (ride.status == DRide.NEW) $ throwError (RideInvalidStatus $ show ride.status)
+  unless (booking.status == DRB.TRIP_ASSIGNED || isInitiatedByCronJob) $ throwError (BookingInvalidStatus $ show booking.status)
```
Comment (jaypalmudaliyar24) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/Common.hs:502]: Not required for `RIDE_STARTED`.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/Common.hs [hunk 3]
```diff
@@ -516,7 +517,7 @@ validateDriverArrivedReq DriverArrivedReq {..} = do
   let BookingDetails {..} = bookingDetails
   booking <- runInReplica $ QRB.findByBPPBookingId bppBookingId >>= fromMaybeM (BookingDoesNotExist $ "BppBookingId: " <> bppBookingId.getId)
   ride <- QRide.findByBPPRideId bppRideId >>= fromMaybeM (RideDoesNotExist $ "BppRideId" <> bppRideId.getId)
-  unless (isValidRideStatus ride.status) $ throwError $ RideInvalidStatus "The ride has already started."
+  unless (isValidRideStatus ride.status || isInitiatedByCronJob) $ throwError $ RideInvalidStatus "The ride has already started."
```
Comment (jaypalmudaliyar24) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/Common.hs:520]: Not required for Driver arrived event.

**Review by jaypalmudaliyar24 — APPROVED** [review]

File: Backend/dev/migrations-read-only/dynamic-offer-driver-app/ride.sql [hunk 4]
```diff
@@ -54,4 +54,14 @@ ALTER TABLE atlas_driver_offer_bpp.ride ADD COLUMN toll_charges double precision
...
+
+
+------- SQL updates -------
+
+ALTER TABLE atlas_driver_offer_bpp.ride ADD COLUMN ride_ended_by text ;
+
+
+------- SQL updates -------
+
+ALTER TABLE atlas_driver_offer_bpp.ride ALTER COLUMN merchant_operating_city_id SET NOT NULL;
```
[the remark this answers is not in the document]
Comment (max-keviv) [reply Backend/dev/migrations-read-only/dynamic-offer-driver-app/ride.sql:67]: it is generated by DSL, it was not present in previous pr

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/Common.hs (hunk 1 above)
Comment (max-keviv) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/Common.hs:483]: can you please explain why it is not required in the RIDE_ASSIGNED validation ?

Comment (jaypalmudaliyar24) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/Common.hs]: CronJob only tries to bring ride and booking to terminal state, this events don't lead to terminal state.

**Review by jaypalmudaliyar24 — APPROVED** [review]

**Review by NikithShetty — APPROVED** [review]

**Pull request merged**